com.unity.render-pipelines.high-definition/Editor/Material/BRGPicking.hlsl

```diff
@@ -0,0 +1,131 @@
+#include "Packages/com.unity.render-pipelines.core/ShaderLibrary/UnityInstancing.hlsl"
+
+struct PickingAttributesMesh
+{
+    float3 positionOS   : POSITION;
+    UNITY_VERTEX_INPUT_INSTANCE_ID
+};
+
+struct PickingMeshToDS
+{
+    float3 positionRWS : INTERNALTESSPOS;
+    UNITY_VERTEX_INPUT_INSTANCE_ID
+};
+
+struct PickingMeshToPS
+{
+    float4 positionCS : SV_Position;
+    UNITY_VERTEX_INPUT_INSTANCE_ID
+};
+
+#ifdef TESSELLATION_ON
+#define PickingVertexOutput PickingMeshToDS
+#else
+#define PickingVertexOutput PickingMeshToPS
+#endif
+
+float4x4 unity_BRGPickingViewMatrix;
+float4x4 unity_BRGPickingProjMatrix;
+float4 unity_BRGPickingCameraWorldPos;
+
+#undef unity_ObjectToWorld
+float4x4 LoadObjectToWorldMatrixBRGPicking()
+{
+    float4x4 objectToWorld = LoadDOTSInstancedData_float4x4_from_float3x4(UNITY_DOTS_INSTANCED_METADATA_NAME(float3x4, unity_ObjectToWorld));
+#if (SHADEROPTIONS_CAMERA_RELATIVE_RENDERING != 0)
+    objectToWorld._m03_m13_m23 -= unity_BRGPickingCameraWorldPos.xyz;
+ #endif
+    return objectToWorld;
+}
+
+float4 ComputePositionCS(float3 positionWS)
+{
+    float4x4 viewMatrix = unity_BRGPickingViewMatrix;
+    // HDRP expects no translation in the matrix because of camera relative rendering
+    viewMatrix._m03_m13_m23_m33 = float4(0,0,0,1);
+    return mul(unity_BRGPickingProjMatrix, mul(viewMatrix, float4(positionWS, 1)));
+}
+
+PickingVertexOutput Vert(PickingAttributesMesh input)
+{
+    PickingVertexOutput output;
+    ZERO_INITIALIZE(PickingVertexOutput, output);
+    UNITY_SETUP_INSTANCE_ID(input);
+    UNITY_TRANSFER_INSTANCE_ID(input, output);
+
+    float4x4 objectToWorld = LoadObjectToWorldMatrixBRGPicking();
+    float4 positionWS = mul(objectToWorld, float4(input.positionOS, 1.0));
+
+#ifdef TESSELLATION_ON
+    output.positionRWS = positionWS;
+#else
+    output.positionCS = ComputePositionCS(positionWS.xyz);
+#endif
+
+    return output;
+}
+
+// No-op Tessellation stage to make at least the shader compile correctly
+#ifdef TESSELLATION_ON
+
+#define MAX_TESSELLATION_FACTORS 1.0
+
+struct PickingTessellationFactors
+{
+    float edge[3] : SV_TessFactor;
+    float inside : SV_InsideTessFactor;
+};
+
+PickingTessellationFactors HullConstant(InputPatch<PickingMeshToDS, 3> input)
```

Comment:
I don't get it, if you define your own tessellation pass, then it will not work with shader graph controlled tessellation. (haven't read the full PR yet, so maybe there is something later). My guess is that this picking pass must be define for all material. Same thing for vertex deformation. Have you tested with it?

Reply:
The picking isn't intended to work with displaced triangles. This is only needed to make sure the shader compiles. For example, the picking pass from the LitTessellation shader declares `#pragma hull Hull` and `#pragma domain Domain`, which means that the tessellation shader stages must be defined at some point. That's what `BRGPicking.hlsl` was doing by just defining passthrough tessellation stages. Although all of this doesn't apply to the new iteration since now it uses an override picking shader instead of a fallback. Now there is no need for any tessellation stage anymore because it can simply be omitted entirely.